### PR TITLE
Add a way for users to enter a license key

### PR DESCRIPTION
**Pull request by poisa**
This PR adds a way for end users to be able to set a license key into Jetpack. Hosting partners and agencies have their own docs where they are instructed to directly set the correct WP option themselves. This new UI opens up that possibility to end users. This is crucial at this point in order to support hosting partners using the WHMCS (pbtFPC-Qd-p2) integration, otherwise they'll get a licenses from their hosting partner but will not be able to do anything with it.

Given the time constraints and team member availability this UI has been massively scoped down to the bare minimum. Currently as the user submits the license key Jetpack will set the correct license option. There is already code in Jetpack that listens to the update event on the license and talks to WPCOM to do what's needed to set the correct Jetpack plan based on the license. So the scope of this PR is only to show the UI and to set the option, that's it.

#### Jetpack product discussion

Project thread: pbtFPC-11l-p2

#### Does this pull request change what data or activity we track or use?

No. It only adds a way to set a wp option for the license key.

#### Testing instructions:

* Note: there is no need to have a valid Jetpack license to test this PR.
* Go to the My Plan page and check that the UI is **not** there
* Add the Code Snippets plugin and add and enable this snippet `add_filter( 'jetpack_licensing_ui_enabled', '__return_true' );`
* Go back to the My Plan page and check it looks like this:

Screen Shot 2021-02-22 at 18 54 05

* Enter a license (or any text) and a successful toast should appear on the top right. 
* Check the `jetpack_licenses` option (use whatever method you like) and the text you entered should be in there. If you're using the Docker setup you can run `yarn docker:wp option get jetpack_licenses` and you should see an array printed out with the license you entered earlier. If you entered more than one license then all of them should be shown there.

#### Proposed changelog entry for your changes:

* Add UI to set license key.

### CODE REVIEW

**Review by zinigor — APPROVED** [review]
I can confirm that this PR works as expected - the license gets saved to an option. I haven't tested it with actual real licenses, though.

File: projects/plugins/jetpack/_inc/lib/class.core-rest-api-endpoints.php
Comment (brbrr) [inline projects/plugins/jetpack/_inc/lib/class.core-rest-api-endpoints.php:3992]: `update_option` might fail to `append_license`. Not sure if we want to return a non-200 response in that case.

File: projects/plugins/jetpack/_inc/client/my-plan/my-plan-header/license.jsx
Comment (jeherve) [inline projects/plugins/jetpack/_inc/client/my-plan/my-plan-header/license.jsx:16]: We try to stay away from classes for new components, and now use function components instead:
https://github.com/Automattic/jetpack/blob/e13487d0dbbe4c743739b6f1a3e37c143a8c2614/projects/plugins/jetpack/_inc/client/README.md#L8-L10

File: projects/plugins/jetpack/_inc/client/my-plan/my-plan-header/license.jsx
Comment (jeherve) [inline projects/plugins/jetpack/_inc/client/my-plan/my-plan-header/license.jsx:5]: Instead of relying on `apiFetch`, could we follow an approach similar to the other elements of the dashboard? You can check this example for previous art on this: #17011 (see `updateLicensingError` to see how things are handled).

File: projects/plugins/jetpack/_inc/lib/class.core-rest-api-endpoints.php
Comment (jeherve) [inline projects/plugins/jetpack/_inc/lib/class.core-rest-api-endpoints.php:685]: Should this be accessible to everyone who has access to the admin page? This would include a site's subscribers. Maybe checking for the `jetpack_connect_user` capability would be more appropriate?

File: projects/plugins/jetpack/_inc/client/my-plan/my-plan-header/index.js
Comment (jeherve) [inline projects/plugins/jetpack/_inc/client/my-plan/my-plan-header/index.js:266]: Should we show this to all Jetpack site owners? Given that only a subset of sites will be able to use licenses, it may be confusing if we add this input field here. It may generate support requests of folks asking for a license when they cannot actually get one.

Would it be an option to only display the field when a filter is set on the site, and that filter would then be set in an mu-plugin by a partner supporting licensing, kind of like we did with the Wizard setup?

**Review by jeherve — CHANGES_REQUESTED** [review]

File: projects/plugins/jetpack/_inc/lib/class.core-rest-api-endpoints.php
Comment (IanRamosC) [reply projects/plugins/jetpack/_inc/lib/class.core-rest-api-endpoints.php]: I've added a custom permission callback `set_jetpack_license_key_permission_check`. It only checks for `jetpack_connect_user`, similar to some other permission callbacks, but I added this as a separate callback so we can change the condition specifically for that route if needed.

File: projects/plugins/jetpack/_inc/client/my-plan/my-plan-header/index.js
Comment (IanRamosC) [reply projects/plugins/jetpack/_inc/client/my-plan/my-plan-header/index.js]: I added the code to only display this part of the UI if the filter `jetpack_show_licensing_ui` is `true`. I'd like to get feedback on my approach here.

Comment (jeherve) [reply projects/plugins/jetpack/_inc/client/my-plan/my-plan-header/index.js]: Yeah, I think that works well that way! 👍

**Comment by IanRamosC** [comment]
I've addressed the feedback and I'm going to request a review from my team.

**Review by atanas-dev — COMMENTED** [review]
Tests good for me - just 2 minor comments.

File: projects/plugins/jetpack/_inc/client/my-plan/my-plan-header/license.jsx
Comment (IanRamosC) [reply projects/plugins/jetpack/_inc/client/my-plan/my-plan-header/license.jsx:51]: Yeah, they don't change and my initial thought was to not add them as deps, but the linter raises a warning (I could disable it for this line, but I'd rather add them as dependencies, considering my ~lack of~ familiarity with the code base).

**Review by atanas-dev — APPROVED** [review]
Tests good for me 👍

File: projects/packages/licensing/src/class-licensing.php
Comment (jeherve) [inline projects/packages/licensing/src/class-licensing.php:242]: Could you add a docblock here, so our parser can pick up the filter and automatically create a page in the Jetpack Codex for that new filter? Here is an example:
https://github.com/Automattic/jetpack/commit/0118cad3cc73ba0eba7635d88b1f2e9618cd75f2

**Review by jeherve — CHANGES_REQUESTED** [review]

File: projects/packages/licensing/src/class-licensing.php
Comment (IanRamosC) [reply projects/packages/licensing/src/class-licensing.php:242]: Thanks for the suggestion! I've added the docblock.

Comment (poisa) [reply projects/packages/licensing/src/class-licensing.php:242]: @IanRamosC Did you forget to push this commit? 🤔

**Comment by poisa** [comment]
@jeherve looks like this is all good from our end. If you folks aren't too busy with HACK week we'd love to get a final review on this so that we can mark our task as done. Thanks!

File: projects/plugins/jetpack/_inc/client/my-plan/my-plan-header/license.jsx
Comment (jeherve) [reply projects/plugins/jetpack/_inc/client/my-plan/my-plan-header/license.jsx:45]: Do you think we could modify a bit the UI so site owners can see when a license has been set? I'm afraid that if we don't, we'll cause some confusion. Even if we refactor / redesign that interface in the future, the version that we'll release in 9.6.0 will be available on some users' sites and may stay there if they don't update. I'd consequently prefer if it were as usable as possible.

I just created #19135 as a bit of a proposal you could maybe start from, or let me know what you think?

Comment (poisa) [reply projects/plugins/jetpack/_inc/client/my-plan/my-plan-header/license.jsx:45]: I don't think this may be entirely possible to achieve. Right now the endpoint that Jetpack calls to get its available  plans only gets store subscriptions. There is no indication that says whether any subscriptions are linked to a license. There is ongoing work to eventually be able to do this but we are still at least a month away from that if not more.

The only thing we could possibly know right now is if the option has one or more licenses (what you are checking in #19135). These licenses may or may not be valid. Even if they are there is no way to tell if they have been used on some other site so we can't really consider what's in the option as the source of truth. @atanas-dev do you think that when the attach process fails we can remove that license from the option? So at least for now, only licenses that we know that _at some point_ were valid will remain in there and we have _something_ to show?

Having said that, this option can very easily get out of sync which I think will generate a lot more confusion than the current UI. For example, a partner/user puts a license in there. Then an HE or partner revokes the license. At this point the option is showing stale (and invalid) data and we have no way of updating it. Also, this option can hold an array of licenses which may look very strange as a list of comma-separated values and it may not be immediately obvious to a user that to add a new license they'd have to append it to the list.

I do understand that right now the process is less than ideal but this is just until the work in pbtFPC-1ek-p2 gets done. Exposure to this UI will be very very minimal.

Comment (jeherve) [reply projects/plugins/jetpack/_inc/client/my-plan/my-plan-header/license.jsx:45]: > I do understand that right now the process is less than ideal but this is just until the work in pbtFPC-1ek-p2 gets done. Exposure to this UI will be very very minimal.

I cannot say I'm a fan of delivering something that's not finished to customers, but I understand. I assume it wouldn't be an option to have no UI for now, and only instructions for customers to add a code snippet to their site to add a license via a filter, for example? A bit like what we do for affiliates and partner IDs today?

Comment (ebinnion) [reply projects/plugins/jetpack/_inc/client/my-plan/my-plan-header/license.jsx:45]: > I assume it wouldn't be an option to have no UI for now, and only instructions for customers to add a code snippet to their site to add a license via a filter, for example? A bit like what we do for affiliates and partner IDs today?

This already exists and is in use for Jurassic Ninja sites as far as I understand. What we're aiming to provide is something basic that simplifies things even further.

> I cannot say I'm a fan of delivering something that's not finished to customers

Since this UI is hidden behind a filter, which is effectively a feature flag, I would lean towards us shipping and being clear with the agencies and hosting partners that are interested in this feature that it is not complete and that we may ask them to disable it if it causes a lot of confusion and/or support burden. Also, through this filter, we could work with agencies and hosting partners to conditionally allow the UI on certain version of Jetpack as well.

**Review by jeherve — DISMISSED** [review]
Following our discussion in https://github.com/Automattic/jetpack/pull/18905#discussion_r591601450 , this  should now be good to merge.

**Review by jeherve — DISMISSED** [review]
Re-approving after adding a changelog entry, in 4a9a4eb159e90e2f04c4a520b064d3ff2647ba94.
To find out more about how you can add changelog entries to your PRs, you can check p1HpG7-bqa-p2

**Comment by atanas-dev** [comment]
@jeherve could you please look through the changes again - I had to adjust some composer.json-s and want to be sure I've done it properly.

**Review by jeherve — APPROVED** [review]

**Pull request merged**
